workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/WorkflowAndroidXSupport.kt

```diff
@@ -34,10 +38,52 @@ public object WorkflowAndroidXSupport {
    * is found in the view tree.
    */
   @WorkflowUiExperimentalApi
-  public fun stateRegistryOwnerFromViewTreeOrContext(view: View): SavedStateRegistryOwner? =
+  public fun stateRegistryOwnerFromViewTreeOrContext(view: View): SavedStateRegistryOwner =
+    checkNotNull(stateRegistryOwnerFromViewTreeOrContextOrNull(view)) {
+      "Expected to find a SavedStateRegistryOwner either in a parent view or the Context of $view"
+    }
+
+  /**
+   * Tries to get the parent [SavedStateRegistryOwner] from the current view via
+   * [ViewTreeSavedStateRegistryOwner], if that fails it looks up the context chain for a registry
+   * owner, and if that fails it just returns null. This differs from
+   * [ViewTreeSavedStateRegistryOwner.get] because it will check the [View.getContext] if no owner
+   * is found in the view tree.
+   */
+  @WorkflowUiExperimentalApi
+  private fun stateRegistryOwnerFromViewTreeOrContextOrNull(view: View): SavedStateRegistryOwner? =
     ViewTreeSavedStateRegistryOwner.get(view)
       ?: view.context.ownerOrNull(SavedStateRegistryOwner::class)
 
+  /**
+   * Attempts to generate a unique key for the given [containerView] for registering the view on
+   * a [SavedStateRegistry], typically as the `key` parameter to
+   * [StateRegistryAggregator.attachToParentRegistry][com.squareup.workflow1.ui.androidx.StateRegistryAggregator.attachToParentRegistry].
+   *
+   * The [SavedStateRegistry] API involves defining string keys to associate with state bundles.
+   * These keys must be unique relative to the instance of the registry they are saved in. This
+   * function tries to satisfy that requirement by combining [containerView]'s fully-qualified class
+   * name with both its [view ID][View.getId] and the
+   * [compatibility key][com.squareup.workflow1.ui.Compatible.compatibilityKey] of its rendering.
+   * This method isn't guaranteed to give a unique key, but it should be good enough: If you need to
+   * nest multiple containers of the same type under the same [SavedStateRegistry], just wrap each
+   * container's rendering with a [Named] or give each container view a unique view ID.
+   *
+   * There's a potential issue here where if [containerView]'s ID is changed to something else, then
+   * another container is added with the old ID, the new container will overwrite the old one's
+   * state. Since they'd both be using the same key, [SavedStateRegistry] would throw an exception.
+   * However, as long as the first container is detached before its ID is changed this shouldn't be
+   * a problem.
+   */
+  @OptIn(WorkflowUiExperimentalApi::class)
+  public fun createStateRegistryKeyForContainer(containerView: View): String {
+    val nameSuffix = containerView.getRendering<Any>()
+      ?.let { Compatible.keyFor(it) }?.let { "-$it" }
```

Comment:
If `Compatible.keyFor(it)` returns something it won't have a trailing `"-"` right?

Reply:
Nice catch.

Starting to second guess whether this method is even useful. Zach put it in place before I made the `Compatible.keyFor` usage more consistent -- he originally was doing `is Named<*>` checks. I'm going to try deleting this.